### DISCUSSION TITLE
Support for Usernameless

### DISCUSSION THREAD

**Discussion by fturiot** [MasterKale/SimpleWebAuthn#103] [Troubleshooting]
Hello,
has anyone implemented userless webauthn authentication?

thanks for help
regards

**Comment by akanass** [comment]
It's implemented since PR#97 so all is ready to do it

**Comment by fturiot** [comment]
yes ok bit for registration i have residentkey
but how i get user information?
registration change ?
thanks

**Comment by MasterKale** [comment]
@fturiot There's some work going on right now on the SimpleWebAuthn homepage to document how to set up your server to support Usernameless. In the meantime, here are some bits of info that can help:

## 1. Configure attestation options to require resident key and user verification.

Your options should look something like this:
```javascript
{
  "user": {
    // By nature of the fact that this is usernameless you'll have to specify 
    // something here, even if it's not pretty
    "name": " (Usernameless user created at 2/1/2021 5:53:46 PM)",
    // Generate an otherwise legitimate user ID here, unique from any other
    // user account in your DB
    "id": "IChVc2VybmFtZWxlc3MgdXNlciBjcmVhdGVkIGF0IDIvMS8yMDIxIDU6NTM6NDYgUE0p"
  },
  "authenticatorSelection": {
    "requireResidentKey": true,
    "userVerification": "required"
  }
  // ...all the other options here...
}
```
You will need to generate a random value for `user.id` as usernameless implies that you _can't_ know the user's account ahead of time; essentially, any time a user performs usernameless attestation they'd be creating a new user in your account. 

That's not to say you actually create a new user account in your system before they complete attestation verification; rather, you're generating and temporarily persisting a user ID, as part of the `generateAttestationOptions()` step, that will get applied to the account you create _after_ the user taps their authenticator and your server verifies the response via `verifyAttestationResponse()`.

## 2. For assertions, don't specify anything for `allowCredentials`.

Allow any authenticator with a credential for your RP ID to perform login.

## 3. Reference the user ID that is returned in the assertion response

By nature of the fact that you won't have usernames as part of your system, **your server will have to rely on the user ID returned during assertion to know which user is logging into your system.** This is the `userHandle` value in `AuthenticatorAssertionResponseJSON` (`body.response.userHandle`) that will be populated in the authenticator's response.

Assuming `verifyAssertionResponse()` returns `true` for `verified`, your server would then establish a session for the user using account information row in your DB with the row ID/userHandle column equal to `userHandle`.

---

This is all off the top of my head as I haven't sat down yet and tried to implement usernameless myself. There's some good info in this [How To FIDO doc the FIDO Alliance is working on](https://github.com/fido-alliance/how-to-fido/blob/master/HowToFIDO.md) that might also help. I don't think they yet cover usernameless but there's still some good info on implementing WebAuthn login for at least 2FA and Passwordless.

**Reply by MartinJaskulla** [reply]
I am trying to go usernameless on Mac OS and iOS without an external security key. Is this supported?

When I don't specify anything for allowCredentials as per `2. For assertions, don't specify anything for allowCredentials`,
```js
startAssertion({
      "challenge": "zJTGJ1yTqtL3l4jkDctTqx_7_Im7r0s7_bgshbmKbw8",
      "timeout": 60000,
      "userVerification": "preferred",
      "rpId": "localhost"
})
```
I get this error on Chrome:
![Screenshot 2021-05-31 at 09 39 39](https://user-images.githubusercontent.com/35801656/120159716-c919ed80-c1f5-11eb-9757-ce8ace7b95e9.png)

**Reply by patte** [reply]
Thanks for the write up.

I spottet the following typo in "Your options should look something like this:"
>```
>   "authenticatorSelection": {
>     "requireResidentKey": true,
>     "userVerification": "required"
>   }
>```

Must be the following, due to [AttestationOptions.ts#L122](https://github.com/MasterKale/SimpleWebAuthn/blob/master/packages/server/src/attestation/generateAttestationOptions.ts#L122).
```
  "authenticatorSelection": {
      residentKey: "required",
     "userVerification": "required"
  }
```
Note to myself: use typescript

**Reply by MasterKale** [reply]
@MartinJaskulla That's a desktop Chrome error modal - after some further testing it seems Chrome just doesn't support usernameless. If you try that same `startAssertion()` call in Safari on Big Sur you should get this "saved account" modal that lets you choose an account to log in with:

<img width="461" alt="Screen Shot 2021-05-31 at 8 47 53 PM" src="https://user-images.githubusercontent.com/5166470/120264455-b2ee4a80-c252-11eb-845f-c97a77782b7f.png">

Accounts that show up in this modal are assigned to a resident key because Safari _always_ generates resident key credentials (because as per the WebAuthn spec [authenticators are allowed to generate resident keys for generated credentials even if a resident key isn't explicitly required...](https://www.w3.org/TR/webauthn-2/#sctn-op-make-cred)). Clicking one of these accounts will then submit that credential to the RP, at which point the RP can use the `"userHandle"` property to identify the user that's logging in. That's the usernameless flow in a nutshell; unfortunately it's not feasible with Chrome.

@patte So _technically_ you _can_ set just `"requireResidentKey"`, as that was the original way to require a "resident key" (redefined to "discoverable credential" in L2) in L1 of the spec. The section of code you called out above forces `"requireResidentKey"` to `true` when the new property, `"residentKey"`, was introduced in L2 of the spec and mandated that the older property be set to certain values when `"residentKey"` is specified.

As for typing, TypeScript won't give you any trouble setting both as per the `AuthenticatorSelectionCriteria` interface:

```ts
export interface AuthenticatorSelectionCriteria {
    authenticatorAttachment?: AuthenticatorAttachment;
    requireResidentKey?: boolean;
    residentKey?: ResidentKeyRequirement;
    userVerification?: UserVerificationRequirement;
}
```

I'd agree with you, though, that it's better to use `"residentKey"` going forward.

**Reply by MartinJaskulla** [reply]
Thanks a lot! Can confirm it works in Safari, but not in Chrome and Firefox.

**Reply by shu8** [reply]
> ## 3. Reference the user ID that is returned in the assertion response
> By nature of the fact that you won't have usernames as part of your system, **your server will have to rely on the user ID returned during assertion to know which user is logging into your system.** This is the `userHandle` value in `AuthenticatorAssertionResponseJSON` (`body.response.userHandle`) that will be populated in the authenticator's response.
> 
> Assuming `verifyAssertionResponse()` returns `true` for `verified`, your server would then establish a session for the user using account information row in your DB with the row ID/userHandle column equal to `userHandle`.

@MasterKale Apologies if this is the wrong place for me to ask a question following on from this (please redirect me if so!).

Thanks a lot for building this, it's really great! I'm struggling with just one part of the usernameless flow, namely how to store the challenge created from `generateAuthenticationOptions` (server GET) against the user (who we do not know at this point), so that it can be verified via the `expectedChallenge` in `verifyAuthenticationResponse` (server POST)?

I think I understand that in `verifyAuthenticationResponse` (POST), we would use the `userHandle` to find the Challenge to pass as `expectedChallenge`, but I think I may be missing something obvious for the step before, to actually save the authentication Challenge against the (unknown) user in the first place!

I'd be really grateful if you had any advice regarding this! :) 

(I think my question is similar to https://github.com/MasterKale/SimpleWebAuthn/discussions/103#discussioncomment-834408 but unfortunately I still couldn't quite figure out what to do here.)

**Comment by fturiot** [comment]
yes thanks, but user in this form is not accepted in simpleauthn 
replace by userId and username ? 
i test  
thanks for all

**Comment by MasterKale** [comment]
> yes thanks, but user in this form is not accepted in simpleauthn 
> 
> replace by userId and username ? 

Ah, yes, to populate those two fields you would indeed populate `userID` and `username` when generating attestation options.

**Comment by fturiot** [comment]
yes thanks, i go to developp login 
thanks for all

**Comment by fturiot** [comment]
in login key populate user information ?
thanks

**Comment by MasterKale** [comment]
I'm not sure I understand your question. Can you provide some more context of what you're asking about?

**Comment by fturiot** [comment]
login passwordless and userless 
the user id  is send by resident key ?

**Comment by MasterKale** [comment]
> login passwordless and userless
> the user id is send by resident key ?

Yes, the value you specify as `userID` during registration/attestation should be the same value the authenticator returns as `response.userHandle` during login/assertion.

**Comment by fturiot** [comment]
ok but i use uuid for id user but problem for decode this 
i have by example '6d1eee38-098e-48f5-abad-8e9a8481fac8' id transmit
but i recept mNqPsbV-d2tFtfyaT34TLA

**Comment by fturiot** [comment]
i think i must be decode

**Comment by MasterKale** [comment]
Yep, decode that to buffer and it should work for `userID`.

Also keep in mind [`user.id` is a buffer with a maximum size of 64 bytes](https://www.w3.org/TR/webauthn-2/#dom-publickeycredentialuserentity-id). Your UUID is going to be fine, I figured I'd call it out anyway.

**Comment by fturiot** [comment]
thanks for all
just one last question
in
Verify attestation response
the id populate is it userId ?
thanks for all
regards




-- 
Frédéric TURIOT

**Comment by MasterKale** [comment]
Unfortunately no, it's the Credential ID. You don't get `user.id` back in an attestation response, you have to persist it like you do `challenge`.

**Comment by MasterKale** [comment]
@fturiot I'm going to convert this into a [Discussion](https://github.com/MasterKale/SimpleWebAuthn/discussions) since it's turned into a very interesting conversation I don't want getting lost in the Closed Issues section of the repo.

**Reply by fturiot** [reply]
i have solve to send in header user.id

**Comment by akanass** [comment]
@MasterKale I just wanted to let you know that I ended up implementing authentication without a password or username and everything is working fine. I will publish my demo app and I could give you the link so you can test. I will also provide my GitHub so you can see the code behind. Thank you for this beautiful bookstore which helped me a lot

**Reply by MasterKale** [reply]
That's great news, and thank you for the kind words! I'd love to see how you pulled it off when you're comfortable sharing 😄

**Comment by timstrasser** [comment]
When using resident keys, what "expectedChallenge" would one provide when verifying the assertion response? As username-less implies, we don't know anything about the user that's going to log in, and we can't just use the returned userHandle until we verified the assertion response.

Also, what I don't understand: How can we verify the signature if we don't know the public key?

**Reply by MasterKale** [reply]
Usernameless assertion is exactly the same as 2FA or Passwordless assertion:

1. Call `generateAssertionOptions()`, specifying your own challenge or letting the library generate one, and passing an empty array for `allowCredentials`
2. Store the value of `challenge` in the object returned from Step 1
3. Give the options to `navigator.credentials.get()`
4. Feed the response from Step 3 into `verifyAssertionResponse()`, specifying the challenge you stored in Step 2 as `expectedChallenge`

> As username-less implies, we don't know anything about the user that's going to log in, and we can't just use the returned userHandle until we verified the assertion response.
>
> Also, what I don't understand: How can we verify the signature if we don't know the public key?

Maybe this is where the confusion lies - usernameless authentication still requires a user to have registered, via `"direct"` attestation, with your site! Upon successful registration you associate a credential ID and public key to the random, new user ID you generated and specified in the attestation options. That credential ID and public key are then used during authentication to verify the signature, at which point you can also verify `userHandle` with the user ID on your end to make sure they match (or however you want to proceed once the assertion response is verified) and then consider the user "logged in".

**Reply by timstrasser** [reply]
I'm still a bit confused.
So when registering we're saving credential ID, public key (aren't these the same here?) and (random) user ID in the database. So far so good.

Now for the login:
We send the client a new challenge but as we don't know anything about the user yet, we cannot associate the challenge with any user yet.

So when we receive the assertion response, how can we know the challenge that was signed was actually the challenge that we sent earlier? Is there a proper way of doing that?

**Reply by akanass** [reply]
All is related to the `userHandle` this how to identifie the user. When the user registers on your application, you have to store the `userHandle` with the data of the user then you have to find him by this value when you want to login

**Reply by akanass** [reply]
And for usernameless/passwordless you have to make sure you have `attestationType: direct`, `residentKey: required` & `userVerification: required` for security purpose

**Reply by akanass** [reply]
And for the challenge, you have to store it somewhere (cookie, db, redis...) during the start to retrieve it during the verify

**Reply by timstrasser** [reply]
Thanks, this makes sense. Is there a specific reason attestationType should be "direct"? After all, seting the attestationType to "direct" adds another prompt when the user registers.

**Reply by MasterKale** [reply]
...That's actually a good question, I can't remember why I recommended direct attestation. I don't think that's a requirement, requiring resident keys is the most important part of all this 😓

**Reply by timstrasser** [reply]
Ok, I'll try to implement this soon. Might be an idea to include passwordless / usernameless auth as an example in this repo. After all resident keys seem to be widely used and the example that's currently provided gives, at a first glance, the impression of being password- & usernameless anyways.

**Off Topic:** Does anyone here by chance know how to remove resident keys from a MacBook? I've created many useless accounts and struggle to find a way to remove them again.

**Reply by MasterKale** [reply]
> **Off Topic:** Does anyone here by chance know how to remove resident keys from a MacBook? I've created many useless accounts and struggle to find a way to remove them again.

You can't 💀

![Screen Shot 2021-05-14 at 11 50 31 AM](https://user-images.githubusercontent.com/5166470/121079563-26321800-c78f-11eb-9f34-a8034574c282.png)

**Reply by MasterKale** [reply]
I don't think my Touch ID even stores resident key credentials anymore for my dev domain because I filled up all the slots 😭

**Reply by timstrasser** [reply]
Oof, I was afraid of that. It was already hard enough to find and remove them on Windows but Apple just "stepped it up to the next level". 😂

**Reply by MasterKale** [reply]
I am desperately refreshing the WWDC 2021 Developer site looking for Safari WebAuthn news, hoping they gave macOS a Windows Hello-level API for Touch ID, and Safari the ability to manage resident keys 😂

**Reply by akanass** [reply]
Direct attestation is useful to store the full attestation and validate it with the metadata service from Fido.

I don't have another popup doing this and for a bank/government application or an application which needs the higher level of security it's better to do it.

**Reply by akanass** [reply]
To delete all previous registration you have to go in your browser settings in security section and for an application you can delete all

**Reply by akanass** [reply]
I already have implemented an application with usernameless and passwordless. I will do the documentation to explain how to use it through docker and I will share my repo here like this you will be able to see how is working and how it's implemented because all works fine for me

**Reply by timstrasser** [reply]
> I am desperately refreshing the WWDC 2021 Developer site looking for Safari WebAuthn news, hoping they gave macOS a Windows Hello-level API for Touch ID, and Safari the ability to manage resident keys 😂

I would have expected the credentials to be stored in keychain... would have made so much sense.
I'm still disappointed no new MacBooks Pro were announced.

**Reply by timstrasser** [reply]
> To delete all previous registration you have to go in your browser settings in security section and for an application you can delete all

I'm not sure how I could delete anything from there.
<img width="810" alt="image" src="https://user-images.githubusercontent.com/53019576/121081320-dbd78800-c7dc-11eb-9756-05ba084b7c7f.png">

**Reply by timstrasser** [reply]
> I already have implemented an application with usernameless and passwordless. I will do the documentation to explain how to use it through docker and I will share my repo here like this you will be able to see how is working and how it's implemented because all works fine for me

That's nice. I'd like to see that.

**Reply by timstrasser** [reply]
> Direct attestation is useful to store the full attestation and validate it with the metadata service from Fido.
> 
> I don't have another popup doing this and for a bank/government application or an application which needs the higher level of security it's better to do it.

I think I've seen this additional popup somewhere, not sure where. It tells the user that the website would like to access information about the security key, like the model. Somewhere I read that direct attestation is generally discouraged to protect the users privacy.

But yes, you're right: It will enhance the security.

**Reply by akanass** [reply]
I am not using `Safari` but `Brave` which is the same as `Chrome` and I can delete all my previous `security keys` registration by going to `additional settings - privacy and security - managed security keys`

I don't have any problem with the `TouchID` as well because it is registered as a `password` for the website and I can delete it then it won't appear anymore by going to `additional settings - privacy and security - delete browsing data (password)`

You have to find the same things for `Safari` and you will be able to manage `security keys` and `TouchID` registrations

**Reply by akanass** [reply]
Yes of course a popup explains to the user that the browser will access to the security keys but it's better for security purpose and I am storing all information that I can because security is never enough

**Reply by timstrasser** [reply]
> You have to find the same things for `Safari` and you will be able to manage `security keys` and `TouchID` registrations

I searched for quite some time, I can't imaging where this could be hidden in Safari. On iOS clearing all browser data removes all credentials with it... But that's also very questionable.

Btw. greetings from Luxembourg ;)

**Reply by akanass** [reply]
@timstrasser `TouchID` is only a password so maybe if you go through the `password` tab you will see it.

For `security keys` I don't know where is it because it's a dedicated section in `Brave/Chrome`

**Reply by timstrasser** [reply]
For security keys it's stored on the key itself (for my Yubikey is is atleast). From there I can remove them with the Yubikey Manager CLI.

But for TouchID, it doesn't seem like the saved resident keys appear anywhere. They don't appear in the passwords tab nor do they appear in Keychain Access.

**Reply by akanass** [reply]
@timstrasser yes you can manage through `Yubikey Manager` and in Brave you can manage it directly.

For `TouchID` as I said, you have to delete the browsing data password in `Safari` like in the screenshot for `Brave`.

You have to find how to delete all the browsing data and select only the password. It's like for cookies or images but you should have `password` too

<img width="509" alt="Capture d’écran 2021-06-07 à 22 36 00" src="https://user-images.githubusercontent.com/1673977/121084277-bf3d4f00-c7e0-11eb-85ac-733fcd0e73a6.png">

**Reply by timstrasser** [reply]
Yes, in the "Privacy"-Tab I can "Manage Website Data", from there I can delete it. But unlike on iOS this does not remove the resident keys. In my "Passwords"-Tab I already removed all the passwords.

**Reply by akanass** [reply]
Maybe inside the `privacy` tab and `manage website data`

**Reply by akanass** [reply]
@timstrasser so I don't know because on `Brave/Chome` when I delete `password` from there all is working and I don't have any `TouchID` registration after that

**Reply by timstrasser** [reply]
That window looks like that. Apparently there's only cache that can be removed. But removing it does not remove any credentials.
<img width="833" alt="image" src="https://user-images.githubusercontent.com/53019576/121085063-b26d2b00-c7e1-11eb-9c2a-7df6cb293b80.png">

**Reply by akanass** [reply]
@timstrasser the only way to do it is to delete the password in the browsing data. If `Safari` doesn't allow it you will always have those registrations. You have to find the way to do it and I don't know where is it.

**Reply by timstrasser** [reply]
Might be just impossible, just like @MasterKale said. Let's hope this will change in the future.

**Reply by akanass** [reply]
normally it should be inside the `privacy` tab and `manage website data` and if it's not allowed it's a shame

**Reply by timstrasser** [reply]
I'd have expected it to be located in Keychain Access, just like the other private keys.

**Reply by akanass** [reply]
@timstrasser I don't know if it's working with [this article](https://www.guidingtech.com/delete-saved-passwords-safari-browser-icloud-keychain-mac/)

`TouchID` is just a `password` for the browser itself so it should be stored in the browser data

**Reply by timstrasser** [reply]
Yeah, this article suggests to delete the passwords from the password tab or from the keychain app, but like I said, there isn't any password stored for the website in any of these places...

**Reply by akanass** [reply]
@timstrasser so I'm sorry, I don't know how to do it and the last things is to delete all the history and/or web data

**Reply by timstrasser** [reply]
Already tried that as well... 😂
Easiest thing to do might be to just reinstall MacOS.

**Reply by akanass** [reply]
@timstrasser did you check inside `iCloud Keychain`  because safari uses this one and not the `session` one

**Reply by timstrasser** [reply]
Yes, there's nothing left in there. :/

**Reply by akanass** [reply]
Sorry I can't do more.

I know it's store there in `Brave/Chrome` so I don't know where is it for `Safari`

**Reply by timstrasser** [reply]
No problem, thanks for the effort anyways. :)

**Reply by MasterKale** [reply]
Things might be getting better with macOS Monterey https://developer.apple.com/wwdc21/10106 👀 ![image](https://user-images.githubusercontent.com/5166470/121088255-4d421700-c79a-11eb-8c3f-f753973f7943.png)

**Reply by timstrasser** [reply]
Yaaay! This looks very promising!

**Comment by guillewu** [comment]
Hi all! Very interesting and helpful thread. I was wondering if anyone was able to "combine" register/login? 
The issue I'm bumping into is a bit of a UX issue I guess. The problem is that if the user has registered previously, and somehow they forgot they already registered and they go through the register process again, webauthn will just create a new user. Is there a way to somehow check whether there's already a credential related to the domain the user is trying to log in to and offer them to sign in with that first (before creating another user/credential)?

**Reply by MasterKale** [reply]
Something like this is coming! ["Conditional UI" is a proposed addition to WebAuthn L3](https://github.com/w3c/webauthn/wiki/Explainer:-WebAuthn-Conditional-UI) that would allow RPs to essentially nudge a user to log in with an existing discoverable credential if one exists, without the user needing to initiate an auth attempt that the RP knows will fail. It's an opportunistic API addition specifically because of how hard it is, for example, to detect when the user has already enrolled their computer's platform authenticator without attempting authentication and falling back to registration if that fails.

In a usernameless context there's still a chance of a user re-registering if they jump onto a new device since usernameless necessarily utilizes an empty `allowCredentials` array. In that case you'd need to come up with some kind of bootstrapping mechanism and instruct your users on how to use it if they want to log in from multiple machines 😬

Or perhaps it makes you push your users to use discoverable-credential-capable security keys instead 🤔

**Reply by JonahStein** [reply]
Is there a time table for Conditional UI?

**Reply by MasterKale** [reply]
Not one any of us using this in production should count on. It'll be an addition to WebAuthn L3 (L2 is the current version) and once L3 gets officially published it then becomes a matter of time for browser vendors to actually implement that functionality.

What's likely to happen is that Chrome and Edge Chromium will likely get initial Conditional UI support before L3 lands (they tend to be on the forefront of practical WebAuthn implementation as far as browsers are concerned, often helping drive definition of the spec), then Safari will probably get it, then if we're lucky Firefox gets on board too.

If I had to guess? We're a six months to a year out from Conditional UI becoming available in Chrome, and you'll still need to implement fallback behavior for other browsers that don't support it yet.
